### PR TITLE
test: edgecase only e20 in neighbour grid area

### PR DESCRIPTION
**Pull request by ClausQAEnerginet**
# Description

Covers edge case https://app.zenhub.com/workspaces/mandalorian-5fd22a1a59e4740018f5ab5a/issues/gh/energinet-datahub/team-mandalorian/310

See readme.md for full description. 

Includes some cleanup/reorganizing of existing logic tests

## Pull-request quality

- [X] The title adheres to [this guide](https://github.com/Mech0z/GitHubGuidelines)
- [X] Tests are written and executed locally
- [ ] ~Subsystem tests have been tested (by manually deploying to `dev_002`)~
- [ ] ~Documentation has been updated~
- [ ] ~The integration [event catalog](https://energinet.atlassian.net/wiki/spaces/D3/pages/555581556/Event+catalog) has been updated~
- [ ] ~C4 diagrams have been updated and Team Outlaws has been informed about relevant changes~

### CODE REVIEW

File: source/databricks/calculation_engine/tests/features/utils/Use-template.ps1 [hunk 1]
```diff
@@ -16,11 +16,13 @@ function CheckFileExists
 
 $templatesPath = "templates/"
 
+$s1 = "/" + $templatesPath + "calculation-test-template.py"
+$s2 = "/" + $templatesPath + "public-model-test-template.py"
 # Define the path to the calculation test template file
-$calculationTestTemplateFilePath = Join-Path $PSScriptRoot $templatesPath "calculation-test-template.py"
+$calculationTestTemplateFilePath = Join-Path $PSScriptRoot $s1
```
Comment (energinet-ajw) [inline source/databricks/calculation_engine/tests/features/utils/Use-template.ps1:22]: My latest PR adds this functionality. This might the reason for the conflict.

Comment (ClausQAEnerginet) [reply source/databricks/calculation_engine/tests/features/utils/Use-template.ps1]: Check

File: source/databricks/calculation_engine/tests/features/given_a_wholesale_calculation/daily_tariff_per_ga_co_es/when_with_child_metering_point/test_output.py [hunk 2]
```diff
@@ -29,4 +29,4 @@ def test__equals_expected(
     actual_and_expected: Any,
     output_name: str,
 ) -> None:
-    assert_output(actual_and_expected, output_name)
+    assert_output(actual_and_expected, output_name, True)
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/features/given_a_wholesale_calculation/daily_tariff_per_ga_co_es/when_with_child_metering_point/test_output.py:32]: What does "True" do?

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/Oracle - exchange between 2 other ga.xlsx
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/Oracle - exchange between 2 other ga.xlsx:1]: Should this file be added?

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/readme.md [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+
+- Input period is post May 2023 so that results are quarterly
+- An E20 sending power into the grid area is included, otherwise we cannot calculate grid loss. But no readings are
+  included for it in time series.
+- Oracle Excel-sheet included (Oracle - exchange between 2 other ga.xlsx)
+
+## Coverage
+
+- E20 (energy in)
+- E20 (two other GA)
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/readme.md:16]:
```suggestion
- E20 (exchange energy in)
- E20 (exchange to other GA)
```

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/readme.md [hunk 4]
```diff
@@ -0,0 +1,21 @@
...
+- E17 (flex)
+- E17 (nonprofiled)
+- E17 (glmp)
+- E18
+- E18 (skmp)
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/readme.md:21]:
```suggestion
- E18 (scmp)
```

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/test_output.py [hunk 5]
```diff
@@ -0,0 +1,32 @@
...
+# IMPORTANT:
+# All test files should be identical. This makes changing them cumbersome.
+# So in order to make it easier you can modify the utils/templates/calculation-test-template.py file instead,
+# and then run the power-shell script "Use-Template.ps1" to update all test_output.py files.
+@pytest.mark.parametrize("output_name", get_output_names())
+def test__equals_expected(
+    actual_and_expected: Any,
+    output_name: str,
+) -> None:
+    assert_output(actual_and_expected, output_name, False)
```
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/test_output.py:32]: False here, but true in all other files?

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_minimal_standard_scenario/EnergyResults-when_minimal_standard_calculation.xlsx
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_minimal_standard_scenario/EnergyResults-when_minimal_standard_calculation.xlsx:1]: Is this being added on purpose?

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_minimal_standard_scenario/Oracle - minimal standard calculation.xlsx
Comment (BjarkeMeier) [inline source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_minimal_standard_scenario/Oracle - minimal standard calculation.xlsx:1]: Is this being added on purpose?

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/Oracle - exchange between 2 other ga.xlsx
Comment (BjarkeMeier) [reply source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/Oracle - exchange between 2 other ga.xlsx:1]: What is "Oracle"? And what is in this file?

**Review by BjarkeMeier — COMMENTED** [review]
nit: What is the purpose of the "#" in the beginning of the title?

File: source/databricks/calculation_engine/tests/features/given_a_wholesale_calculation/daily_tariff_per_ga_co_es/when_with_child_metering_point/test_output.py (hunk 2 above)
Comment (ClausQAEnerginet) [reply source/databricks/calculation_engine/tests/features/given_a_wholesale_calculation/daily_tariff_per_ga_co_es/when_with_child_metering_point/test_output.py:32]: Parameter that only shows failing columns (similar to the way we only show mismatching rows)

Comment (BjarkeMeier) [reply source/databricks/calculation_engine/tests/features/given_a_wholesale_calculation/daily_tariff_per_ga_co_es/when_with_child_metering_point/test_output.py:32]: It's always hard to guess the semantics of boolean flags.
So this would be a nice improvement:
```suggestion
    assert_output(actual_and_expected, output_name, show_only_mismatching_rows=True)
```

Comment (ClausQAEnerginet) [reply source/databricks/calculation_engine/tests/features/given_a_wholesale_calculation/daily_tariff_per_ga_co_es/when_with_child_metering_point/test_output.py:32]: I don't disagree, but it's one for Alex/the framework, I guess. For what it's worth, the name of the parameter does show up when you view the file in Pycharm

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/Oracle - exchange between 2 other ga.xlsx
Comment (ClausQAEnerginet) [reply source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/Oracle - exchange between 2 other ga.xlsx:1]: About the Excel-sheets. "Oracle" is a testing term for a resource or artifact used for verifying expected results. 
The Excel-sheet added here (and for the other tests) is the mini-calculator I made for myself to generate results for the different calculations. It contains the same MP's (with "stamdata" (engelsk?)) and quantities used in the test so that I 
1. Don't have to accept the calculation results at face value
2. Don't have to calculate them myself by hand with rounding and other business
If there is a mismatch, I can easily compare the input data with the Excel-sheet and see where I forgot to add a value etc. 
I included it along with the test, because if I have to modify the test, I have easy access to the sheet that was used to generate the results, if I want to e.g. add an MP to the input data or change quantities and want to see how it affects results. 
I don't think it should be mandatory to have it, but I figured it didn't hurt either. I can clarify the existence of these sheets when I get around to updating the documentation for the framework.

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_minimal_standard_scenario/Oracle - minimal standard calculation.xlsx
Comment (ClausQAEnerginet) [reply source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_minimal_standard_scenario/Oracle - minimal standard calculation.xlsx:1]: See answer above about the Excel-sheets

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/readme.md (hunk 4 above)
Comment (ClausQAEnerginet) [reply source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/readme.md:21]: Hmm. English vs. Danish. I have not seen it referred to as "scmp" anywhere, but I have also not seen the term "System Correction Metering Point" anywhere. Is "scmp" an established term?

Comment (BjarkeMeier) [reply source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/readme.md:21]: I don't recall exactly when to use negative grid loss and when to use system correction grid loss...
But we should neither way use a danish abbreviation. Btw, do you expect the reader to know scmp/skmp and glmp?

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/Oracle - exchange between 2 other ga.xlsx
Comment (BjarkeMeier) [reply source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/Oracle - exchange between 2 other ga.xlsx:1]: Interesting 😄 
I think it would be nice to inform the team. Others might even want to use it as well.

File: source/databricks/calculation_engine/tests/features/given_a_wholesale_calculation/daily_tariff_per_ga_co_es/when_with_child_metering_point/test_output.py (hunk 2 above)
Comment (BjarkeMeier) [reply source/databricks/calculation_engine/tests/features/given_a_wholesale_calculation/daily_tariff_per_ga_co_es/when_with_child_metering_point/test_output.py:32]: PyCharm is one thing. Reviewing and reading code is (often) another thing.
@energinet-ajw 👀

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_minimal_standard_scenario/EnergyResults-when_minimal_standard_calculation.xlsx
Comment (BjarkeMeier) [reply source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_minimal_standard_scenario/EnergyResults-when_minimal_standard_calculation.xlsx:1]: My bad.

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/Oracle - exchange between 2 other ga.xlsx
Comment (ClausQAEnerginet) [reply source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/Oracle - exchange between 2 other ga.xlsx:1]: I will put it on the framework documentation page when I get around to it (when done with energy edge cases)

File: source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/readme.md (hunk 4 above)
Comment (ClausQAEnerginet) [reply source/databricks/calculation_engine/tests/features/given_an_energy_calculation/when_exchange_between_two_other_ga/readme.md:21]: I guess I assumed that they were known abreviations. I use them to avoid writing out "Systemkorrektionsmålepunkt"/"System Correction Metering Point". 
I'll start a discussion thread with the rest of the team and closing the conversation here.

**Review by energinet-ajw — APPROVED** [review]
